libexec/rtld-elf/rtld_c18n.h

```diff
@@ -118,9 +118,34 @@ struct stk_table {
 
 #define	COMPART_ID_MAX		index_to_cid((stk_table_index) { -1 })
 
+#include "rtld_c18n_machdep.h"
+
 struct tcb *c18n_allocate_tcb(struct tcb *);
 void c18n_free_tcb(void);
 
+/*
+ * When entering the RTLD without a trampoline (e.g., during lazy binding, TLS
+ * lookup, or stack resolution), a dummy trusted frame indicating that the
+ * current compartment is RTLD must be pushed.
+ */
+static inline struct trusted_frame *
+push_dummy_rtld_trusted_frame(struct trusted_frame *tf)
+{
+	*--tf = (struct trusted_frame) {
+		.callee = cid_to_index(RTLD_COMPART_ID)
+	};
+	set_trusted_stk(tf);
+	return (tf);
+}
+
+static inline struct trusted_frame *
+pop_dummy_rtld_trusted_frame(struct trusted_frame *tf)
+{
+	assert(get_trusted_stk() == tf);
```

Comment:
So why not just make it implicit?...

Reply:
I want to retain the assertion because it would be very hard to debug if something goes wrong here.

Reply:
What does asserting that the input is the one valid value achieve? The function can just use get_trusted_stk() itself, it doesn't need an argument.

Reply:
My main worry is that somebody pushed to the trusted stack and forgot to pop. The assertion helps to catch that.

Reply:
And how does this code help catch that?

Reply:
As in, this assertion checks that nobodies made unbalanced push/pops to the trusted stack between the pair of push and pop that I am doing right now.